### PR TITLE
feat(compiler): add full directive metadata and validation logic

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/angular2/src/compiler/api.ts [hunk 1]
```diff
@@ -12,6 +13,27 @@ export class TypeMeta {
   }
 }
 
+export class ChangeDetectionMeta {
```
Comment (gkalpak) [inline modules/angular2/src/compiler/api.ts]: Why not call this (and `TemplateMeta`) ...MEta**data** for consistency ?

Comment (pkozlowski-opensource) [inline modules/angular2/src/compiler/api.ts]: Agree with @gkalpak - it would a longer, but a more explicit name.

File: modules/angular2/src/compiler/api.ts [hunk 2]
```diff
@@ -12,6 +13,27 @@ export class TypeMeta {
...
+  }) {
+    this.changeDetection = changeDetection;
+    this.properties = properties;
+    this.events = events;
+    this.hostListeners = hostListeners;
+    this.hostProperties = hostProperties;
+  }
+}
+
 export class TemplateMeta {
```
Comment (pkozlowski-opensource) [inline modules/angular2/src/compiler/api.ts]: Same here: `TemplateMeta` -> `TemplateMetadata`

File: modules/angular2/src/core/render/dom/schema/dom_element_schema_registry.ts [hunk 3]
```diff
@@ -1,17 +1,28 @@
...
+    return element;
+  }
+
+  hasProperty(tagName: string, propName: string): boolean {
     if (tagName.indexOf('-') !== -1) {
       // can't tell now as we don't know which properties a custom element will get
       // once it is instantiated
       return true;
     } else {
+      var elm = this._getProtoElement(tagName);
```
Comment (pkozlowski-opensource) [inline modules/angular2/src/core/render/dom/schema/dom_element_schema_registry.ts]: Now that we create instances of elements, I think that it should work with web components out of the box. So we could probably simplify this method end remove the if / else branch.

But we can definitively do it in a separate PR with tests etc.

File: modules/angular2/src/compiler/api.ts [hunk 4]
```diff
@@ -12,6 +13,27 @@ export class TypeMeta {
   }
 }
 
+export class ChangeDetectionMeta {
+  changeDetection: string;
```
Comment (tbosch) [inline modules/angular2/src/compiler/api.ts]: is an enum now

Comment (tbosch) [inline modules/angular2/src/compiler/api.ts]: remove it for now.

File: modules/angular2/src/compiler/api.ts [hunk 5]
```diff
@@ -54,18 +76,24 @@ export enum ViewEncapsulation {
...
   selector: string;
-  constructor({type, selector}: {type?: TypeMeta, selector?: string} = {}) {
+  hostAttributes: Map<string, string>;
+  changeDetection: ChangeDetectionMeta;
+  template: TemplateMeta;
+  constructor({type, isComponent, selector, hostAttributes, changeDetection, template}: {
+    type?: TypeMeta,
+    isComponent?: boolean,
+    selector?: string,
+    hostAttributes?: Map<string, string>,
```
Comment (tbosch) [inline modules/angular2/src/compiler/api.ts]: remove this, not used yet.

File: modules/angular2/src/compiler/api.ts [hunk 6]
```diff
@@ -54,18 +76,24 @@ export enum ViewEncapsulation {
...
+    changeDetection?: ChangeDetectionMeta,
+    template?: TemplateMeta
+  } = {}) {
     this.type = type;
+    this.isComponent = isPresent(isComponent) ? isComponent : false;
```
Comment (tbosch) [inline modules/angular2/src/compiler/api.ts]: use `lang/normalizeBool`

**Comment by tbosch** [comment]
Add a try/catch for expression parser and store the error in the `TemplateVisitor.errors` as well

**Comment by tbosch** [comment]
Reviewed with @vsavkin in person
